packages/notifier/src/publish-kit.js

```diff
@@ -180,12 +200,221 @@ export const makePublishKit = () => {
+      advanceCurrent(true, undefined, makeQuietRejection(reason));
     },
   });
   return harden({ publisher, subscriber });
 };
 harden(makePublishKit);
+
+// TODO: Move durable publish kit to a new file?
+
+/**
+ * @typedef {ReturnType<typeof initDurablePublishKitState>} DurablePublishKitState
+ */
+
+/**
+ * @param {DurablePublishKitOptions} [options]
+ */
+const initDurablePublishKitState = (options = {}) => {
+  const { valueDurability = 'mandatory' } = options;
+  assert.equal(valueDurability, 'mandatory');
+  return {
+    // configuration
+    valueDurability,
+
+    // lifecycle progress
+    publishCount: 0n,
+    status: 'live', // | 'finished' | 'failed'
+
+    // persisted result data
+    hasValue: false,
+    value: undefined,
+  };
+};
+
+/** @type {WeakMap<object, {currentP, tailP, tailR}>} */
+const durablePublishKitEphemeralData = new WeakMap();
+
+// XXX Does state actually work as a WeakMap key?
+/**
+ * Returns the current/next-result promises and next-result resolver
+ * associated with a given state object backing a durable publish kit.
+ * They are lost on upgrade, but recreated on-demand.
+ * Such recreation preserves the value in (but not the identity of) the
+ * current { value, done } result when possible, which is always the
+ * case when that value is terminal (i.e., from `finish` or `fail`) or
+ * when the durable publish kit is configured with
+ * `valueDurability: 'mandatory'`.
+ *
+ * @param {DurablePublishKitState} state
+ */
+const provideDurablePublishKitEphemeralData = state => {
+  const foundData = durablePublishKitEphemeralData.get(state);
+  if (foundData) {
+    return foundData;
+  }
+  const { status, publishCount } = state;
+  let newData;
+  if (status === 'failed') {
+    newData = {
+      currentP: makeQuietRejection(state.value),
+      tailP: makeQuietRejection(
+        new Error('Cannot read past end of iteration.'),
+      ),
+      tailR: undefined,
+    };
+  } else if (status === 'finished') {
+    // XXX Should we be harden()ing rejection-reason errors?
+    const tailP = makeQuietRejection(
+      new Error('Cannot read past end of iteration.'),
+    );
+    newData = {
+      currentP: HandledPromise.resolve(
+        harden({
+          head: { value: state.value, done: true },
+          publishCount,
+          tail: tailP,
+        }),
+      ),
+      tailP,
+      tailR: undefined,
+    };
+  } else if (status === 'live') {
+    const { promise: tailP, resolve: tailR } = makePromiseKit();
+    newData = {
+      currentP: state.hasValue
+        ? HandledPromise.resolve(
+            harden({
+              head: { value: state.value, done: false },
+              publishCount,
+              tail: tailP,
+            }),
+          )
+        : tailP,
+      tailP,
+      tailR,
+    };
+  } else {
+    assert.fail(X`Invalid durable promise kit status: ${q(status)}`);
+  }
+  durablePublishKitEphemeralData.set(state, harden(newData));
+  return newData;
```

Comment:
```suggestion
  return harden(newData);
```
or harden each in place. Or dispense with the `newData` variable and harden and return each in place. FWIW, that's what I'd do.